File: web/src/lib/api/spread.test.ts
import { afterEach, describe, expect, it, vi } from "vitest";

import { fetchBasisSpreadHistory, mapBasisSpreadHistory } from "./spread";

function historyWire(overrides: Record<string, unknown> = {}) {
  return {
    venue: "binance",
    baseAsset: "BTC",
    quoteAsset: "USDT",
    canonicalSymbol: "BTCUSDT",
    range: "24h",
    resolutionSeconds: 60,
    availability: "available",
    asOf: "2026-08-22T07:00:00Z",
    points: [
      {
        ts: "2026-08-22T06:59:00Z",
        spreadBps: "12.5",
        spotAsk: "100",
        perpetualAsk: "100.125",
        samples: 2,
      },
    ],
    summary: {
      currentBps: "12.5",
      minBps: "8",
      maxBps: "15",
      avgBps: "11",
      coverage: "0.97",
    },
    ...overrides,
  };
}

afterEach(() => {
  vi.unstubAllGlobals();
});

describe("basis spread mapping", () => {
  it("maps decimal strings without treating them as percents", () => {
    const history = mapBasisSpreadHistory(historyWire());
    expect(history.points[0]?.spreadBps).toBe(12.5);
    expect(history.summary.coverage).toBe(0.97);
    expect(history.availability).toBe("available");
  });
});

describe("basis spread fetch", () => {
  it("returns unchanged without parsing a 304 body", async () => {
    const fetchMock = vi.fn().mockResolvedValue(
      new Response(null, {
        status: 304,
        headers: { ETag: '"spread-1"' },
      }),
    );
    vi.stubGlobal("fetch", fetchMock);
    await expect(
      fetchBasisSpreadHistory("Binance", "btc", "usdt", "24h", '"spread-1"'),
    ).resolves.toEqual({
      status: "unchanged",
      etag: '"spread-1"',
    });
    expect(fetchMock).toHaveBeenCalledWith(
      "/api/v1/basis-spreads/binance/BTC/USDT/history?range=24h",
      expect.objectContaining({
        headers: {
          Accept: "application/json",
          "If-None-Match": '"spread-1"',
        },
      }),
    );
  });
});

File: backend/internal/rpc/spread_test.go
package rpc

import (
	"context"
	"testing"
	"time"

	shopspring "github.com/shopspring/decimal"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
	spreadv1 "selfquant/backend/gen/spread/v1"
	"selfquant/backend/internal/spread"
)

type stubSpreadService struct {
	last    spread.HistoryRequest
	history spread.History
	err     error
}

func (s *stubSpreadService) GetHistory(
	_ context.Context,
	request spread.HistoryRequest,
) (spread.History, error) {
	s.last = request
	return s.history, s.err
}

func TestSpreadServerMapsHistory(t *testing.T) {
	now := time.Date(2026, 8, 22, 7, 0, 0, 0, time.UTC)
	server := NewSpreadServer(&stubSpreadService{history: spread.History{
		Venue: "binance", BaseAsset: "BTC", QuoteAsset: "USDT",
		CanonicalSymbol: "BTCUSDT", Range: spread.Range1h, ResolutionSeconds: 5,
		Availability: spread.AvailabilityAvailable, AsOf: now,
		Points: []spread.Point{{
			TS: now, SpreadBps: shopspring.RequireFromString("12.5"),
			SpotAsk: shopspring.RequireFromString("100"), PerpetualAsk: shopspring.RequireFromString("100.125"),
			Samples: 2,
		}},
		Summary: spread.Summary{
			CurrentBps: shopspring.RequireFromString("12.5"),
			MinBps:     shopspring.RequireFromString("12.5"),
			MaxBps:     shopspring.RequireFromString("12.5"),
			AvgBps:     shopspring.RequireFromString("12.5"),
			Coverage:   shopspring.RequireFromString("0.01"),
		},
	}})
	response, err := server.GetBasisSpreadHistory(context.Background(), &spreadv1.GetBasisSpreadHistoryRequest{
		Venue: "binance", BaseAsset: "BTC", QuoteAsset: "USDT",
		Range: spreadv1.BasisSpreadRange_BASIS_SPREAD_RANGE_1H,
	})
	if err != nil {
		t.Fatal(err)
	}
	if response.GetAvailability() != spreadv1.BasisSpreadAvailability_BASIS_SPREAD_AVAILABILITY_AVAILABLE ||
		response.GetCanonicalSymbol() != "BTCUSDT" ||
		len(response.GetPoints()) != 1 ||
		response.GetPoints()[0].GetSpreadBps() != "12.5" {
		t.Fatalf("response=%v", response)
	}
}

func TestSpreadServerForwardsCompareVenue(t *testing.T) {
	stub := &stubSpreadService{history: spread.History{
		Venue: "binance", CompareVenue: "okx", BaseAsset: "BTC", QuoteAsset: "USDT",
		CanonicalSymbol: "BTCUSDT", Range: spread.Range24h, ResolutionSeconds: 60,
		Availability: spread.AvailabilityAvailable, AsOf: time.Date(2026, 8, 22, 7, 0, 0, 0, time.UTC),
	}}
	server := NewSpreadServer(stub)
	response, err := server.GetBasisSpreadHistory(context.Background(), &spreadv1.GetBasisSpreadHistoryRequest{
		Venue: "binance", CompareVenue: "okx", BaseAsset: "BTC", QuoteAsset: "USDT",
		Range: spreadv1.BasisSpreadRange_BASIS_SPREAD_RANGE_24H,
	})
	if err != nil {
		t.Fatal(err)
	}
	if stub.last.CompareVenue != "okx" || response.GetCompareVenue() != "okx" {
		t.Fatalf("last=%+v response=%v", stub.last, response)
	}
}

func TestSpreadServerRejectsMissingRange(t *testing.T) {
	server := NewSpreadServer(&stubSpreadService{})
	_, err := server.GetBasisSpreadHistory(context.Background(), &spreadv1.GetBasisSpreadHistoryRequest{
		Venue: "binance", BaseAsset: "BTC", QuoteAsset: "USDT",
	})
	if status.Code(err) != codes.InvalidArgument {
		t.Fatalf("err=%v", err)
	}
}

func TestMapSpreadError(t *testing.T) {
	if status.Code(mapSpreadError(spread.ErrInvalidArgument)) != codes.InvalidArgument {
		t.Fatal("invalid argument mapping")
	}
	if status.Code(mapSpreadError(spread.ErrTimeout)) != codes.DeadlineExceeded {
		t.Fatal("timeout mapping")
	}
	if status.Code(mapSpreadError(spread.ErrQueryFailed)) != codes.Unavailable {
		t.Fatal("query mapping")
	}
}
